### PR TITLE
[wip] Add a Client.Provider that will go through a chain of providers.

### PR DESCRIPTION
**Pull request by keichan34**
This is still a work in progress; most of it is minimally functional, but I wanted to share my work to see if this is a direction you want to proceed with.

Here's a rough to-do list for what I want to see in this PR.
- [x] Getting credentials from environment variables
- [ ] Getting credentials from the `~/.aws/config` and `~/.aws/credentials` files
- [x] Getting credentials from EC2 instance metadata (IAM role)
- [ ] EC2 instance metadata failure retry modes

### CODE REVIEW

File: lib/aws/client.ex [hunk 1]
```diff
@@ -5,7 +5,27 @@ defmodule AWS.Client do
...
-            endpoint: nil,
-            service: nil
+            endpoint: "amazonaws.com",
+            service: nil,
+            expires_at: nil
+
+  @type t :: %AWS.Client{}
+
+  @spec merge(AWS.Client.t, AWS.Client.t) :: AWS.Client.t
+  def merge(a, b) do
```
Comment (sinetris) [inline lib/aws/client.ex]: Can be simplified as:
```elixir
def merge(a, b) do
  Map.merge a, b, fn
    (_k, v1, nil) ->
      v1
    (_k, _v1, v2) ->
      v2
  end
end
````
:wink:

Comment (keichan34) [inline lib/aws/client.ex]: Heh, thanks! I knew there must be some way to pattern match on anonymous functions! :+1:

**Comment by jkakar** [comment]
@keichan34 Sorry for never responding to this; somehow I missed it until now. :/  In the meantime, not knowing about this, I've created https://github.com/jkakar/aws-erlang-metadata to solve the instance metadata problem.

**Comment by jkakar** [comment]
I'm going to close this out because it's old and out-of-date.  Please feel free to reopen it if you'd prefer to keep it active.

**Pull request closed without merging**